transformers no longer ignore UnApply.fun

xeno-by opened the pull request:
Second time's the charm. I remember trying to do exactly the same somewhen
around 2.10.0-M4, but then some continuations tests were failing.
Luckily, today everything went smoothly.

Please note that this fix changes the way that SI-5465 manifests itself.
Previously it produced type errors, now it simply crashes the compiler.
Therefore I had to attach the try/catch FatalError clause to invocations
of toolbox methods, so that compiler crashes get caught and translated to
ToolBoxErrors.

Also fixes SI-7871, and that clears the way for implementing quasiquotes
with conventional macros rather than relying on a special case in typer.

xeno-by committed:
transformers no longer ignore UnApply.fun

Second time's the charm. I remember trying to do exactly the same somewhen
around 2.10.0-M4, but then some continuations tests were failing.
Luckily, today everything went smoothly.

Please note that this fix changes the way that SI-5465 manifests itself.
Previously it produced type errors, now it simply crashes the compiler.
Therefore I had to attach the try/catch FatalError clause to invocations
of toolbox methods, so that compiler crashes get caught and translated to
ToolBoxErrors.

Also fixes SI-7871, and that clears the way for implementing quasiquotes
with conventional macros rather than relying on a special case in typer.

xeno-by commented:
review @retronym

retronym commented:
Maybe its a bit outside the scope of this patch, but seeing as though we're in the area, I would prefer this sort of design:

```
object withCompiler {
  private lazy val global = { ... }
  private def cleanup() { ... }
  def apply[A](f: ToolBoxGlobal => A): A = try {
    f(global)
  } catch {
     case _: FatalError => ... 
  } finally {
     cleanup()
  }
}

...
withCompiler { compiler =>

}
```

I'm assuming it's cheap and correct to do the cleanup in all cases, even if not strictly necessary like in `parse`.

This way you can't forget to use the wrappers.

xeno-by committed:
refactors ToolBoxGlobal wrappers

Centralizes wrappers in a single location and makes it impossible to use
the compiler without going through that location.

Also fixes the option string tokenization bug that was there for ages.
This time I figured out that there's already an implementation of the
tokenizer in our compiler, so I just used it :)

xeno-by commented:
@retronym Done

retronym commented on src/compiler/scala/tools/reflect/ToolBoxFactory.scala:
```diff
@@ -28,6 +31,13 @@ abstract class ToolBoxFactory[U <: JavaUniverse](val u: U) { factorySelf =>
 
+    lazy val arguments = CommandLineParser.tokenize(options)
+    lazy val virtualDirectory =
+      arguments.iterator.sliding(2).collectFirst{ case Seq("-d", dir) => dir }.lastOption match {
```
`collectFirst` returns `Option`

xeno-by replied:
Hahaha

retronym commented on src/compiler/scala/tools/reflect/ToolBoxFactory.scala:
```diff
@@ -28,6 +31,13 @@ abstract class ToolBoxFactory[U <: JavaUniverse](val u: U) { factorySelf =>
+    lazy val arguments = CommandLineParser.tokenize(options)
+    lazy val virtualDirectory =
+      arguments.iterator.sliding(2).collectFirst{ case Seq("-d", dir) => dir }.lastOption match {
+        case Some(outDir) => scala.tools.nsc.io.AbstractFile.getDirectory(outDir)
```
Add an import for this for balance with `VirtualDirectory`.

retronym commented:
LGTM
